### PR TITLE
Pipe: Optimize lock in PipeTsFileResourceManager using segment lock strategy

### PR DESCRIPTION
**Pull request by SteveYurongSu**
As title.

### CODE REVIEW

**Comment by SteveYurongSu** [comment]
Refer: 
- https://github.com/apache/iotdb/pull/13888
- https://github.com/apache/iotdb/pull/13891

Thanks @Caideyipi.

**Review by DanielWang2035 — APPROVED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/tsfile/PipeTsFileResourceManager.java [hunk 1]
```diff
@@ -132,42 +131,53 @@ private void ttlCheck() {
...
-          isTsFile
-              ? FileUtils.createHardLink(file, hardlinkOrCopiedFile)
-              : FileUtils.copyFile(file, hardlinkOrCopiedFile);
+    // If the file is a tsfile, create a hardlink in pipe dir and will return it.
+    // otherwise, copy the file (.mod or .resource) to pipe dir and will return it.
+    final File resultFile =
+        isTsFile
+            ? FileUtils.createHardLink(file, hardlinkOrCopiedFile)
+            : FileUtils.copyFile(file, hardlinkOrCopiedFile);
+    segmentLock.lock(resultFile);
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/tsfile/PipeTsFileResourceManager.java:165]: Race issue...

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/tsfile/PipeTsFileResourceManager.java [hunk 2]
```diff
@@ -132,42 +131,53 @@ private void ttlCheck() {
...
+    try {
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/tsfile/PipeTsFileResourceManager.java:166]: in lock

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/tsfile/PipeTsFileResourceManager.java [hunk 3]
```diff
@@ -132,42 +131,53 @@ private void ttlCheck() {
...
-      // If the file is not a hardlink or copied file, check if there is a related hardlink or
-      // copied file in pipe dir. if so, increase reference count and return it
-      final File hardlinkOrCopiedFile = getHardlinkOrCopiedFileInPipeDir(file);
-      if (increaseReferenceIfExists(hardlinkOrCopiedFile.getPath())) {
+    // If the file is not a hardlink or copied file, check if there is a related hardlink or
+    // copied file in pipe dir. if so, increase reference count and return it
+    final File hardlinkOrCopiedFile = getHardlinkOrCopiedFileInPipeDir(file);
+    segmentLock.lock(hardlinkOrCopiedFile);
+    try {
+      if (increaseReferenceIfExists(hardlinkOrCopiedFile)) {
```
Comment (luoluoyuyu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/resource/tsfile/PipeTsFileResourceManager.java:150]: I think `getHardlinkOrCopiedFileInPipeDir(file)` and `segmentLock.lock(hardlinkOrCopiedFile)`; this part of the code can be completed in the `file lock`. Similarly, `resultFile` is obtained in the `hardlinkOrCopiedFile lock`, which can better reduce lock contention.

**Review by luoluoyuyu — APPROVED** [review]

**Pull request merged**